client/src/main/java/io/pravega/client/security/auth/JwtTokenProviderImpl.java

```diff
@@ -232,13 +232,14 @@ String refreshToken() {
         } else {
             log.debug("Token is already under refresh for scope {} and stream {}", this.scopeName, this.streamName);
         }
-        try {
-            currentRefreshFuture.join(); // Block until the token is refreshed
-        } finally {
-            this.tokenRefreshFuture.compareAndSet(currentRefreshFuture, null); // Token is already refreshed, so resetting the future to null.
-        }
-        LoggerHelpers.traceLeave(log, "refreshToken", traceEnterId, this.scopeName, this.streamName);
-        return delegationToken.get().getValue();
+
+        final CompletableFuture<Void> handleToCurrentRefreshFuture  = currentRefreshFuture;
+        return currentRefreshFuture.thenApply(v -> {
```

Comment:
This block could be moved up to inside the if on line 231 so that it is only called once.

Reply:
We do want to run this block in either case (`if`/`else`). So, I can't move it inside the `if`. Am I missing something?

Reply:
I don't think we do.
When a token expires there may be multiple concurrent invocations of `refreshToken`. In this case the first one will reset the token future on line 231. Subsequent calls will take the else branch and end up chaining off the same future. This is correct.
But consider in this case how many times line 239 is executed. It will be run once for each call waiting on the future. Which could be many times. However as this is doing a compare an set, it only makes sense to call this once. Similarly the log line on 240, it only makes sense for it to happen once, when the token is refreshed, not once for every task that was blocked on the future. The only other line is 241 which is mapping the void to the token itself. This is needlessly indirect. The future returned from recreateToken can just be made to return the token buy adding a `return token;` on 247.

Reply:
Since the `whenComplete` takes a `BiConsumer` as input, I cannot have it return the `DelegationToken` object. I can do something like the following instead, but it'll change the existing logic a great deal (the implications of which needs to be studied) in `refreshToken() and elsewhere in the class`. 

```
private CompletableFuture<String> recreateToken() {
        return controllerClient.getOrRefreshDelegationTokenFor(scopeName, streamName)
                .whenComplete((token, ex) -> this.delegationToken.set(new DelegationToken(token, extractExpirationTime(token))));
    }
```
Regarding your other points: 
* The exit trace log is actually intended for every call to `refreshToken`, whether it returns a new delegation token created by the current thread or one created by another thread. 
* Similarly, the block below the if/else condition will return a new delegation token whether it was created by the current thread or created by another thread. 

In any case, I have an [enhancement issue](https://github.com/pravega/pravega/issues/4358) to look at optimizations in this code. We can look at improvements holistically as part of that issue. What do you say?